chain/manager.go

```diff
@@ -298,6 +301,55 @@ func (m *Manager) AddBlocks(blocks []types.Block) error {
 	return nil
 }
 
+// AddValidatedV2Blocks ingests a chain of v2 blocks. The blocks must already be
+// validated, and the first block's parent must be known. If the chain has
+// sufficient work, it may become the new best chain, triggering a reorg.
+func (m *Manager) AddValidatedV2Blocks(blocks []types.Block, states []consensus.State) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if len(blocks) == 0 {
+		return nil
+	}
+	if _, ok := m.store.State(blocks[0].ParentID); !ok {
+		return fmt.Errorf("missing parent for block %v", blocks[0].ParentID)
+	}
+	for i := range blocks {
+		if blocks[i].V2 == nil {
+			return fmt.Errorf("only v2 blocks can be pre-validated")
+		}
+		m.store.AddBlock(blocks[i], &consensus.V1BlockSupplement{})
+		m.store.AddState(states[i])
+	}
+
+	// if this chain is now the best chain, trigger a reorg
+	cs := states[len(states)-1]
+	if cs.SufficientlyHeavierThan(m.tipState) {
```

Comment:
Everything below this line can probably be extracted into a `triggerReorg` helper.

Reply:
eh, it gets kinda ugly due to the onReorg functions needed to be called without a lock; `triggerReorg` would be a weird function that assumes it is called under lock, but then unlocks in the middle? Plus the name ambiguity of `m.reorgTo` vs `m.triggerReorg`. Not worth it imo